### PR TITLE
Put execution to failed state on validation error

### LINKED ISSUE #1042: Execution output validation should put the execution in failed
The error of execution output validation should put the execution in failed state instead of being returned to the service.

At the same time, the `execution.Completed` and `execution.Failed` functions should be merged to one function `execution.Process` (or better name) that take the decision to put the execution in completed or failed state based on the data returned by the service.

Related function:
https://github.com/mesg-foundation/engine/blob/9c92b9596443674d899d4154d514d82a5039e7ff/sdk/sdk.go#L212-L253

### PR DESCRIPTION
**Pull request by krhubert**
fix #1042

### CODE REVIEW

File: sdk/execution/execution.go [hunk 1]
```diff
@@ -81,6 +81,9 @@ func (e *Execution) processExecution(executionHash hash.Hash, outputData []byte,
 	} else {
 		o, err := e.validateExecutionOutput(exec.InstanceHash, exec.TaskKey, outputData)
 		if err != nil {
+			if err1 := exec.Failed(reterr); err1 != nil {
```
[the remark this answers is not in the document]
Comment (krhubert) [reply sdk/execution/execution.go]: Fixed

**Comment by krhubert** [comment]
fix #1042 

> Because there is a part to merge the completed and failed functions.

Yes there is but is unclear.

>At the same time, the execution.Completed and execution.Failed functions should be merged to one function execution.Process

The `processExecution` is the method that takes the decision. I don't know what `execution.Process` should be. If the processExecution should be moved to `execution.Process` then I'm against it, because it would mean that all logic and dependencies from `validateExecutionOutput` should be moved as well. In that case basic type like Execution would require instance and service sdks as deps, which will cause import cycle:

```
can't load package: import cycle not allowed
package github.com/mesg-foundation/engine/execution
	imports github.com/mesg-foundation/engine/sdk/instance
	imports github.com/mesg-foundation/engine/database
	imports github.com/mesg-foundation/engine/execution
```

**Comment by antho1404** [comment]
You don't need to import the sdk, you just need to resolve the service and pass it to the `execution.Process`. This way execution has a dependency on the service package only which is fine. And this way `execution.Process` has all information to check the data and set the state as valid or invalid and like that it simplifies everything

**Comment by krhubert** [comment]
> This way execution has a dependency on the service package only which is fine. 

I don't think it's fine. Execution shouldn't have `service` as a dependency. I think we can extract part about validators/validation from service and use it inside service and execution. In that case, they share the only part that is needed for both packages.

> like that it simplifies everything

It is all the same but the functions are moved from sdk to execution package, but everything stays the same.

I think we merge this PR because it solves the most important issues and we can talk about refactoring in different issue/pr.

**Comment by antho1404** [comment]
It doesn't work. I tried to create a service that publishes invalid outputs and I have nothing on my stream of execution, my task is pending and I have an error logs in the engine logs.

How did you test that? what does it solve?

image

File: sdk/execution/execution.go [hunk 2]
```diff
@@ -81,11 +81,10 @@ func (e *Execution) processExecution(executionHash hash.Hash, outputData []byte,
 	} else {
 		o, err := e.validateExecutionOutput(exec.InstanceHash, exec.TaskKey, outputData)
 		if err != nil {
-			tx.Discard()
-			return nil, err
-		}
-
-		if err := exec.Complete(o); err != nil {
+			if err1 := exec.Failed(err); err1 != nil {
+				err = fmt.Errorf("%s: %s", err, err1)
```
Comment (antho1404) [inline sdk/execution/execution.go]: this error never discards the transaction

**Pull request merged**